lru.go

```diff
@@ -167,7 +167,13 @@ func (lru *LRU[K, V]) hashToBucketPos(hash uint32) uint32 {
 	if lru.mask != 0 {
 		return hash & lru.mask
 	}
-	return hash % lru.size
+	return fastModulo(hash, lru.size)
+}
+
+// fastModulo calculates x % n without using the modulo operator (~4x faster).
+// Reference: https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/
+func fastModulo(x, n uint32) uint32 {
```

Comment:
I'm trying to reproduce this optimization to get the regular modulo operation ~4 times faster. Code at https://gist.github.com/florianl/61de8e1434a513a1beb3f1f76c31cf08

But my benchmarks don't show a significant diference:
```
$ go test -bench=. -benchmem
goos: linux
goarch: amd64
pkg: xch.gl/modulo
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkModulo/FastModulo_3_over_4095-20         	1000000000	         1.141 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/FastModulo_4095_over_3-20         	1000000000	         1.138 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/FastModulo_65535_over_19-20       	1000000000	         1.137 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/FastModulo_19_over_65535-20       	1000000000	         1.134 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/FastModulo_65534_over_19-20       	1000000000	         1.136 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/FastModulo_19_over_65534-20       	1000000000	         1.150 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/FastModulo_4294967294_over_65535-20         	1000000000	         1.161 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/FastModulo_4294967294_over_65534-20         	1000000000	         1.166 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/RegularModulo_3_over_4095-20                	908406850	         1.307 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/RegularModulo_4095_over_3-20                	914586667	         1.298 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/RegularModulo_65535_over_19-20              	912191052	         1.307 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/RegularModulo_19_over_65535-20              	917458092	         1.298 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/RegularModulo_65534_over_19-20              	907173954	         1.310 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/RegularModulo_19_over_65534-20              	915687758	         1.294 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/RegularModulo_4294967294_over_65535-20      	927088612	         1.310 ns/op	       0 B/op	       0 allocs/op
BenchmarkModulo/RegularModulo_4294967294_over_65534-20      	914863524	         1.289 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	xch.gl/modulo	20.744s
```

Reply:
Thanks for benchmarking!

The reason why your benchmarks results look roughly the same is that you are measuring an extremely fast operation. In this case, the overhead of the benchmark code itself becomes much more significant then what you are actually going to measure.

I tried to reduce the overhead by rearranging the benchmark code. Now it shows ~2x improvement.
```
BenchmarkFastModulo-20          1000000000               0.1220 ns/op
BenchmarkRegularModulo-20       1000000000               0.2394 ns/op
```

My assumption is that without the loop overhead, ~4x is in range. The result depends on CPU architecture and model of course. Older CPUs may even be much worse in terms of divisions (modulo is a division operation).
And another aspect is using a small set of fixed values as you did. A good optimizing compiler could special case the data to avoid divisions (though, the Go compiler is possibly not doing things like this). I used a set of random x and n to avoid this, just to be on the safe side.
```
package modulo

import (
	"math"
	"math/rand/v2"
	"testing"
)

func fastModulo(x, n uint32) uint32 {
	return uint32((uint64(x) * uint64(n)) >> 32)
}

func regularModulo(x, n uint32) uint32 {
	return x % n
}

type param struct {
	x, n uint32
}

func BenchmarkFastModulo(b *testing.B) {
	params := make([]param, 0, 128)
	for range params {
		params = append(params, param{x: rand.Uint32(), n: rand.Uint32N(math.MaxUint32-1) + 1})
	}

	for range b.N {
		for i := range params {
			_ = fastModulo(params[i].x, params[i].n)
		}
	}
}

func BenchmarkRegularModulo(b *testing.B) {
	params := make([]param, 0, 128)
	for range params {
		params = append(params, param{x: rand.Uint32(), n: rand.Uint32N(math.MaxUint32-1) + 1})
	}

	for range b.N {
		for i := range params {
			_ = regularModulo(params[i].x, params[i].n)
		}
	}
}
```

Reply:
I had an issue in the benchmark code, fixed it and updated the measured results.

Reply:
Napkin math: On a single 4GHz core a single cycle takes ~0.25 nanoseconds. 
With instruction reordering and executing instruction in parallel CPU cores can bring down this number even further. But everything below 0.2 nanoseconds per operation doesn't sound solid to me.

I'm wondering which benchmark comes closer to the actual use of the functionality. As the modulo operation is just the return operation for some function,  https://gist.github.com/florianl/61de8e1434a513a1beb3f1f76c31cf08 looks to me more closer to the real world use.
But regardless - whether it comes down to 1.141 ns/op vs 1.307 ns/op or 0.1220 ns/op vs 0.2394 ns/op, does the complexity justify the performance improvements?

Reply:
I ran the same tests with an unrolled loop and just 10 params - and the difference is now ~2x. That is much less than the 4x stated in the [blog post](https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/) from 2016.
My guess is that modern x86 CPUs have a faster division today (in comparison to multiplication).

If someone wants to test this on their CPU (`go test ./... -bench=.`):
```
package modulo

import (
	"math"
	"math/rand/v2"
	"testing"
)

func fastModulo(x, n uint32) uint32 {
	return uint32((uint64(x) * uint64(n)) >> 32)
}

func regularModulo(x, n uint32) uint32 {
	return x % n
}

type param struct {
	x, n uint32
}

func makeParams(n int) []param {
	params := make([]param, 0, n)
	for range cap(params) {
		params = append(params, param{x: rand.Uint32(), n: rand.Uint32N(math.MaxUint32-1) + 1})
	}
	return params
}

func BenchmarkFastModulo(b *testing.B) {
	params := makeParams(10)

	for range b.N {
		_ = fastModulo(params[0].x, params[0].n)
		_ = fastModulo(params[1].x, params[1].n)
		_ = fastModulo(params[2].x, params[2].n)
		_ = fastModulo(params[3].x, params[3].n)
		_ = fastModulo(params[4].x, params[4].n)
		_ = fastModulo(params[5].x, params[5].n)
		_ = fastModulo(params[6].x, params[6].n)
		_ = fastModulo(params[7].x, params[7].n)
		_ = fastModulo(params[8].x, params[8].n)
		_ = fastModulo(params[9].x, params[9].n)
	}
}

func BenchmarkRegularModulo(b *testing.B) {
	params := makeParams(10)

	for range b.N {
		_ = regularModulo(params[0].x, params[0].n)
		_ = regularModulo(params[1].x, params[1].n)
		_ = regularModulo(params[2].x, params[2].n)
		_ = regularModulo(params[3].x, params[3].n)
		_ = regularModulo(params[4].x, params[4].n)
		_ = regularModulo(params[5].x, params[5].n)
		_ = regularModulo(params[6].x, params[6].n)
		_ = regularModulo(params[7].x, params[7].n)
		_ = regularModulo(params[8].x, params[8].n)
		_ = regularModulo(params[9].x, params[9].n)
	}
}
```

Reply:
> does the complexity justify the performance improvements?

Well, it is a one-liner (a multiplication and a shift), nothing complex about it.
The [Dolt](https://github.com/dolthub/dolt) database (written in Go) is using the same algorithm for optimization in their [SwissTable implementation](https://github.com/dolthub/swiss/).
They explicitly mention this in a [blog post](https://www.dolthub.com/blog/2023-03-28-swiss-map/) (search for `fastModN`).
For me, these are good reason to also use it.

Note: Testing the SwissTable implementation for FreeLRU is already on my list ;)